lasagne/layers/normalization.py

```diff
@@ -373,3 +375,144 @@ def batch_norm(layer, **kwargs):
     return layer
+
+
+class InstanceNormLayer(Layer):
+    """
+    lasagne.layers.InstanceNormLayer(incoming, axes='auto', epsilon=1e-4,
+     **kwargs)
+    Instance Normalization
+    This layer implements instance normalization of its inputs, following [1]_:
+    .. math::
+        y_i = \\frac{x_i - \\mu_i}{\\sqrt{\\sigma_i^2 + \\epsilon}}
+    That is, each example (instance) in the input is normalized to zero mean
+    and unit variance. The crucial part is that the mean and variance are
+    NOT computed across the batch dimension, i.e., not over examples,
+    but per example. During both training and testing, :math:`\\mu_i` and
+    :math:`\\sigma_i^2` are defined to be the mean and variance
+    of the instances :math:`i` in the current input mini-batch :math:`x`.
+    The advantages of using this implementation over e.g.
+    :class:`BatchNormLayer` with adapted axes arguments, are its
+    independence of the input size, as no parameters are learned and stored.
+
+    Parameters
+    ----------
+    incoming : a :class:`Layer` instance or a tuple
+        The layer feeding into this layer, or the expected input shape
+    axes : 'auto', int or tuple of int
+        The axis or axes to normalize over. If ``'auto'`` (the default),
+        normalize over all axes except for the first two:
+        this will normalize over all spatial dimensions for
+        convolutional layers.
+    epsilon : scalar
+        Small constant :math:`\\epsilon` added to the variance before taking
+        the square root and dividing by it, to avoid numerical problems
+    **kwargs
+        Any additional keyword arguments are passed to the :class:`Layer`
+        superclass.
+    Notes
+    -----
+    This layer should be inserted between a linear transformation (such as a
+    :class:`DenseLayer`, or :class:`Conv2DLayer`) and its nonlinearity. The
+    convenience function :func:`instance_norm` modifies an existing layer to
+    insert instance normalization in front of its nonlinearity.
+    The behavior can be controlled by passing keyword arguments to
+    :func:`lasagne.layers.get_output()` when building the output expression
+    of any network containing this layer.
+    During training, [1]_ normalize each instance in an input mini-batch
+    by its statistics.
+    See also
+    --------
+    batch_norm : Convenience function to apply batch normalization to a layer
+    References
+    ----------
+    .. [1] Ulyanov, D., Vedaldi, A., & Lempitsky, V. (2016):
+           Instance Normalization: The Missing Ingredient for Fast Stylization.
+           https://arxiv.org/pdf/1607.08022.pdf.
+    """
+    def __init__(self, incoming, axes='auto', epsilon=1e-4, **kwargs):
+        super(InstanceNormLayer, self).__init__(incoming, **kwargs)
+
+        if axes == 'auto':
+            # default: normalize over spatial dimensions only,
+            # i.e. separate for each instance in the batch
+            axes = (2, 3)
```

Comment:
You don't know if these axes are there. Should be `tuple(range(2, len(self.input_shape)))`.